ontopy/excelparser.py

```diff
@@ -29,118 +35,248 @@ def create_ontology_from_excel(  # pylint: disable=too-many-arguments
+def create_ontology_from_pandas(  # pylint:disable=too-many-locals,too-many-branches,too-many-statements,too-many-arguments
     data: pd.DataFrame,
     metadata: pd.DataFrame,
+    imports: list,
     base_iri: str = "http://emmo.info/emmo/domain/onto#",
     base_iri_from_metadata: bool = True,
     catalog: dict = None,
+    force: bool = False,
 ) -> Tuple[ontopy.ontology.Ontology, dict]:
     """
     Create an ontology from a pandas DataFrame.
     """
 
-    # Remove Concepts without prefLabel and make all to string
+    # Remove Concepts without prefLabel and make data to string
     data = data[data["prefLabel"].notna()]
-    data = data.astype({"prefLabel": "str"})
+    data = data.astype(str)
+    data["prefLabel"] = data["prefLabel"].str.strip()
+    data = data[data["prefLabel"].str.len() > 0]
+    data.reset_index(drop=True, inplace=True)
 
     # Make new ontology
-    onto, catalog = get_metadata_from_dataframe(metadata, base_iri)
+
+    onto, catalog = get_metadata_from_dataframe(
+        metadata, base_iri, imports=imports
+    )
 
     # base_iri from metadata if it exists and base_iri_from_metadata
     if not base_iri_from_metadata:
         onto.base_iri = base_iri
 
+    labels = set(data["prefLabel"])
+    for altlabel in data["altLabel"].str.strip():
+        if not altlabel == "nan":
+            labels.update(altlabel.split(";"))
+
     onto.sync_python_names()
     with onto:
-        # loop through the rows until no more are added
-        new_loop = True
-        final_loop = False
-        while new_loop:
-            number_of_added_classes = 0
-            for _, row in data.iterrows():
+        remaining_rows = set(range(len(data)))
+        while remaining_rows:
+            added_rows = set()
+            for index in remaining_rows:
+                row = data.loc[index]
                 name = row["prefLabel"]
                 try:
-                    if isinstance(
-                        onto.get_by_label(name), owlready2.ThingClass
-                    ):
-                        continue
+                    onto.get_by_label(name)
+                except (ValueError, TypeError) as err:
+                    warnings.warn(
+                        f'Ignoring concept "{name}". '
+                        f'The following error was raised: "{err}"'
+                    )
+                    continue
                 except NoSuchLabelError:
                     pass
 
-                parent_names = str(row["subClassOf"]).split(";")
-
-                try:
-                    parents = [onto.get_by_label(pn) for pn in parent_names]
-                except NoSuchLabelError:
-                    if final_loop is True:
-                        parents = owlready2.Thing
-
-                        warnings.warn(
-                            "Missing at least one of the defined parents. "
-                            f"Concept: {name}; Defined parents: {parent_names}"
+                if name in onto:
```

Comment:
Has this not already been tested through the `try/else` above?

Reply:
You are right, I moved it above.